### PR TITLE
Connect mobile app with modules

### PR DESCRIPTION
**Pull request by desfero**
**Technical notes**

To progress further with #4007 we need to connect our app with our modules system (sagas and di). 

This pr gives us building blocks we can use to implement a proper abstraction for app storage requirements. 

Also seems I was able to fix `react-native` workspaces linking problems by following and experimenting with the examples from https://github.com/gorhom/react-monorepo.

**New scripts for the mobile app**
`yarn tsc` - compile typescript (at the moment there is something wrong with `react-native` typings will fix in the separate pr)
`yarn start:ios:release and yarn start:android:release` - convenient wrappers to run in release mode

### CODE REVIEW

File: packages/shared-modules/src/extensions.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { IExtension } from "@neufund/sagas";
+import { Container } from "inversify";
+
+import { INeuModule } from "./types";
+
+/**
+ * Load context modules from module libs into global container
+ */
+const getLoadContextExtension = (container: Container): IExtension => ({
+  onModuleAdded: (module: INeuModule<unknown>) => {
```
Comment (DmitriyNoa) [inline packages/shared-modules/src/extensions.ts]: This and onModuleRemoved repeats many times with almost the same signature. Maybe we can move it out and reuse this code?

File: packages/shared-modules/src/extensions.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { IExtension } from "@neufund/sagas";
+import { Container } from "inversify";
+
+import { INeuModule } from "./types";
+
+/**
+ * Load context modules from module libs into global container
+ */
+const getLoadContextExtension = (container: Container): IExtension => ({
```
Comment (DmitriyNoa) [inline packages/shared-modules/src/extensions.ts]: As we have discussed before on the FE discussion meeting on Thursday - let's add some TS docs? What do you think?

File: packages/shared-modules/src/utils.ts [hunk 3]
```diff
@@ -1,5 +1,77 @@
...
 
 const createLibSymbol = <T>(name: string) => Symbol(name) as TLibSymbol<T>;
 
-export { createLibSymbol };
+/**
+ * If react-redux connection function receive `undefined` as a `dispatchToProps` method it then pass `dispatch` method to wrapped component.
+ * Often it's not needed as it leads to problems when wrapped component spread all props to any dom element
+ * @param omitDispatch
+ */
+const getDispatchToProps = (omitDispatch: boolean) => (omitDispatch ? () => ({}) : undefined);
```
Comment (DmitriyNoa) [inline packages/shared-modules/src/utils.ts]: Can you explain why do we return anonymous function with empty object?

File: packages/shared-modules/src/utils.ts [hunk 4]
```diff
@@ -1,5 +1,77 @@
...
+    dispatch: (a: GlobalActions | TModuleActionsOrNever<Module>) => void,
+    ownProps: Props,
+  ) => Dispatch;
+  options?: Options<GlobalState, State, Props> & TCustomOptions;
+};
+
+/**
+ * A wrapper around react-redux connect with build in modules support
+ */
+function appConnect<
```
Comment (DmitriyNoa) [inline packages/shared-modules/src/utils.ts]: Please provide some TS docs as it has many parameters

File: packages/wallet/README.md [hunk 5]
```diff
@@ -16,12 +16,11 @@ device** (just for android development).
 After environment setup is completed you can start the app in emulator (`yarn start:ios` or
 `yarn start:android`).
 
-_**Note**: For ios before starting the app for the first time `pod`'s should be installed manually._
+_**Note**: For ios before starting the app for the first time and after any changes to the ios
+native modules `pod`'s should be installed manually._
```
Comment (DmitriyNoa) [inline packages/wallet/README.md]: Typo after a dot _ ?

File: packages/wallet/app/di/setupBindings.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { coreModuleApi, TLibSymbolType } from "@neufund/shared-modules";
+import { Container } from "inversify";
```
Comment (DmitriyNoa) [inline packages/wallet/app/di/setupBindings.ts]: By the way, what is di stands for, in a folder name?

File: packages/wallet/app/modules/init/actions.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { createActionFactory } from "@neufund/shared";
+
+export const initActions = {
+  start: createActionFactory("INIT_START"),
```
Comment (DmitriyNoa) [inline packages/wallet/app/modules/init/actions.ts]: Let's move all string literals to constants, so we can reuse these later

File: packages/wallet/app/components/Landing.tsx [hunk 8]
```diff
@@ -3,14 +3,34 @@ import { Text, Button, View } from "react-native";
...
+}) => {
   const navigation = useNavigation();
 
+  React.useEffect(() => {
+    init();
+  }, []);
+
   return (
     <View>
       <Text>Landing</Text>
```
Comment (DmitriyNoa) [inline packages/wallet/app/components/Landing.tsx]: I know it's only a stub view, but we need to integrate with some internalization/localization library and see if we can reuse localization files from  web

File: packages/wallet/app/store/create.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  return createStore(
+    {
+      extensions: [
+        getLoadContextExtension(context.container),
+        getContextToDepsExtension(appModule, createGlobalDependencies, context),
+        getSagaExtension(context),
+      ],
+    },
+    setupCoreModule({ backendRootUrl: "fsafa" }),
```
Comment (DmitriyNoa) [inline packages/wallet/app/store/create.tsx]: fsafa ? ) In master we already have react-native-config, to pickup env vars, so we can use it to prove the value. Or just add a todo, so we don't forget to clean it

File: packages/wallet/app/store/types.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { DeepReadonly, Values } from "@neufund/shared";
+import { ActionCreatorsMapObject, Reducer, ReducersMapObject } from "redux";
+
+import { appReducers } from "../modules/reducers";
+
+type TReducersMapToState<T extends ReducersMapObject<any, any>> = T extends ReducersMapObject<
```
Comment (DmitriyNoa) [inline packages/wallet/app/store/types.ts]: Is it possible to replace any ?

File: packages/wallet/app/store/utils.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+// helper to use instead of redux connect. It's bound with our app state and it uses dictionary to pass arguments
+import {
+  appConnect as sharedAppConnect,
+  TAppConnectOptions,
+  TModuleSetup,
+} from "@neufund/shared-modules";
+import { InferableComponentEnhancerWithProps } from "react-redux";
+import { TAppGlobalActions, TAppGlobalState } from "./types";
+
+export function appConnect<
```
Comment (DmitriyNoa) [inline packages/wallet/app/store/utils.ts]: same here. Lets DOC all utils

File: packages/shared-modules/src/utils.ts (hunk 4 above)
Comment (desfero) [reply packages/shared-modules/src/utils.ts]: hm, it has just one which is an object

File: packages/wallet/README.md (hunk 5 above)
Comment (desfero) [reply packages/wallet/README.md]: fixed

File: packages/wallet/app/di/setupBindings.ts (hunk 6 above)
Comment (desfero) [reply packages/wallet/app/di/setupBindings.ts]: it stands for `dependency injection`. I'm thinking whether we can move the content to `libs` folder as its storing all services but for now, that's at least consistent with the `web` app

File: packages/wallet/app/components/Landing.tsx (hunk 8 above)
Comment (desfero) [reply packages/wallet/app/components/Landing.tsx]: agree, will add a ticket today to research whether `react-intl` is still an option for react-native project or maybe we need to use something else

File: packages/wallet/app/store/create.tsx (hunk 9 above)
Comment (desfero) [reply packages/wallet/app/store/create.tsx]: good catch, fixed

File: packages/wallet/app/store/types.ts (hunk 10 above)
Comment (desfero) [reply packages/wallet/app/store/types.ts]: Replaced by `object`

File: packages/wallet/app/store/utils.ts (hunk 11 above)
Comment (desfero) [reply packages/wallet/app/store/utils.ts]: fixed

**Comment by DmitriyNoa** [comment]
Great job. + 100500 for adding TSdocs and extensive comments.

- [x] All e2e tests are passing locally.
- [x] App works.

**Pull request merged**